packages/core/database/lib/entity-manager/relations-orderer.js

```diff
@@ -1,6 +1,106 @@
+ *   { id: 2, position: { end: true } }
+ *
+ * Going through the connect array, id 5 has to be connected before id 1,
+ * so the order of id5 = id1 - 1. But the order value of id 1 is unknown.
+ * The only way to know the order of id 1 is to connect it first.
+ *
+ * This function makes sure the relations are connected in the right order:
+ *   { id: 2, position: { end: true } }
+ *   { id: 1, position: { before: 2 } }
+ *   { id: 5, position: { before: 1 } }
+ *
+ */
+const sortConnectArray = (connectArr, initialArr = [], strictSort = true) => {
+  const sortedConnect = [];
+  // Boolean to know if we have to recalculate the order of the relations
+  let needsSorting = false;
+  // Map to validate if relation is already in sortedConnect or DB.
+  const relationInInitialArray = initialArr.reduce((acc, rel) => ({ ...acc, [rel.id]: true }), {});
+  // Map to store the first index where a relation id is connected
+  const mappedRelations = connectArr.reduce((mapper, relation) => {
+    const adjacentRelId = relation.position?.before || relation.position?.after;
+
+    if (!relationInInitialArray[adjacentRelId] && !mapper[adjacentRelId]) {
+      needsSorting = true;
+    }
+
+    // If the relation is already in the array, throw an error
+    if (mapper[relation.id]) {
+      throw new InvalidRelationError(
+        `The relation with id ${relation.id} is already connected. ` +
+          'You cannot connect the same relation twice.'
+      );
+    }
+
+    return {
+      [relation.id]: { ...relation, computed: false },
+      ...mapper,
+    };
+  }, {});
+
+  // If we don't need to sort the connect array, we can return it as is
+  if (!needsSorting) return connectArr;
+
+  // Recursively compute in which order the relation should be connected
+  const computeRelation = (relation, relationsSeenInBranch) => {
+    const adjacentRelId = relation.position?.before || relation.position?.after;
+    const adjacentRelation = mappedRelations[adjacentRelId];
+
+    // If the relation has already been seen in the current branch,
+    // it means there is a circular reference
+    if (relationsSeenInBranch[adjacentRelId]) {
+      throw new InvalidRelationError(
+        'A circular reference was found in the connect array. ' +
+          'One relation is trying to connect before/after another one that is trying to connect before/after it'
+      );
+    }
+
+    // This relation has already been computed
+    if (mappedRelations[relation.id]?.computed) return;
+
+    mappedRelations[relation.id].computed = true;
+
+    // Relation does not have a before or after attribute or is in the initial array
+    if (!adjacentRelId || relationInInitialArray[adjacentRelId]) {
+      sortedConnect.push(relation);
+      return;
+    }
+
+    // Look if id is referenced elsewhere in the array
+    if (mappedRelations[adjacentRelId]) {
+      computeRelation(adjacentRelation, { ...relationsSeenInBranch, [relation.id]: true });
+      sortedConnect.push(relation);
```

Comment:
What should be the result of the following example ?

in DB : 
```
{ id: 5, order: 1 },
{ id: 6, order: 2 },
{ id: 7, order: 3 }, 
```
array to connect:
```
{ id: 40, after: 6 }
{ id: 20, start: true },
{ id: 6, after: 20 },
{ id: 41, after: 6 }
{ id: 9, before: 10 },
{ id: 10, end: true }
```

Reply:
The sorted Array given by the algo seems to be:
```
      { id: 40, after: 6 },
      { id: 20, start: true },
      { id: 6, after: 20 },
      { id: 41, after: 6 },
      { id: 9, before: 10 },
      { id: 10, end: true }
```
and final result :
```
      { init: true, id: 5, order: 1 },
      { init: true, id: 7, order: 3 },
      { id: 40, after: 6, order: 3.5 },
      { id: 20, start: true, order: 3.5 },
      { id: 6, after: 20, order: 3.5 },
      { id: 41, after: 6, order: 3.5 },
      { id: 9, before: 10, order: 3.5 },
      { id: 10, end: true, order: 3.5 }
```
But I think I expected this order no?:
```
{ id: 20 },
{ id: 6 },
{ id: 41 },
{ id: 5 },
{ id: 40 },
{ id: 7 }, 
{ id: 10 },
{ id: 9 },
```

Reply:
just in case, is the array to connect 
```js
{ id: 40, after: 6 }
{ id: 20, start: true },
{ id: 6, after: 20 },
{ id: 41, after: 6 }
{ id: 9, before: 10 },
{ id: 10, end: true }
```

or you are using the position attribute?
```js
      [
        { id: 40, position: { after: 6 } },
        { id: 20, position: { start: true } },
        { id: 6, position: { after: 20 } },
        { id: 41, position: { after: 6 } },
        { id: 9, position: { before: 10 } },
        { id: 10, position: { end: true } },
      ],
```

Using the second one I get the following sorted array:
```js
      [
        { id: 40, position: { after: 6 } },
        { id: 20, position: { start: true } },
        { id: 6, position: { after: 20 } },
        { id: 41, position: { after: 6 } },
        { id: 10, position: { end: true } },
        { id: 9, position: { before: 10 } },
      ],
```

Reply:
I did a mistake in the syntax yes ! I get the same result as you now, thanks :)

Reply:
phew , I got scared for a sec